File: app/src/main/java/bin/mg/main/handlers/FileClickHandlerRegistry.java
package bin.mg.main.handlers;

import android.content.Context;
import android.view.View;
import bin.mg.main.model.FileItem;
import bin.mg.main.model.FileType;
import java.io.File;
import java.util.HashMap;
import java.util.Map;

public class FileClickHandlerRegistry {
    
    private static final FileClickHandlerRegistry INSTANCE = new FileClickHandlerRegistry();
    private final Map<String, FileClickHandler> handlers = new HashMap<>();
    private FileClickHandler defaultHandler;
    
    private FileClickHandlerRegistry() {}
    
    public static FileClickHandlerRegistry getInstance() {
        return INSTANCE;
    }
    
    public void registerHandler(String extension, FileClickHandler handler) {
        handlers.put(extension.toLowerCase(), handler);
    }
    
    public void registerHandler(FileType type, FileClickHandler handler) {
        handlers.put(type.name().toLowerCase(), handler);
    }
    
    public FileClickHandler getHandler(FileItem item) {
        if (item == null) return defaultHandler;
        
        FileClickHandler handler = handlers.get(item.getExtension().toLowerCase());
        if (handler != null) return handler;
        
        FileType type = item.getFileType();
        handler = handlers.get(type.name().toLowerCase());
        if (handler != null) return handler;
        
        return defaultHandler;
    }
    
    public FileClickHandler getHandler(FileItem item, FileClickHandler fallback) {
        FileClickHandler handler = getHandler(item);
        return handler != null ? handler : fallback;
    }
    
    public void setDefaultHandler(FileClickHandler handler) {
        this.defaultHandler = handler;
    }
    
    public void clearHandlers() {
        handlers.clear();
    }
}

File: app/src/main/java/bin/mg/main/handlers/FileClickHandlerRegistry.kt
package bin.mg.main.handlers

import bin.mg.main.model.FileItem
import bin.mg.main.model.FileType

class FileClickHandlerRegistry private constructor() {

    private val handlers = HashMap<String, FileClickHandler>()
    private var defaultHandler: FileClickHandler? = null

    fun registerHandler(extension: String, handler: FileClickHandler) {
        handlers[extension.lowercase()] = handler
    }

    fun registerHandler(type: FileType, handler: FileClickHandler) {
        handlers[type.name.lowercase()] = handler
    }

    fun getHandler(item: FileItem?): FileClickHandler? {
        if (item == null) return defaultHandler

        val handler = handlers[item.getExtension().lowercase()]
        if (handler != null) return handler

        val type = item.fileType
        val typeHandler = handlers[type.name.lowercase()]
        if (typeHandler != null) return typeHandler

        return defaultHandler
    }

    fun getHandler(item: FileItem, fallback: FileClickHandler): FileClickHandler {
        return getHandler(item) ?: fallback
    }

    fun setDefaultHandler(handler: FileClickHandler) {
        defaultHandler = handler
    }

    fun clearHandlers() {
        handlers.clear()
    }

    companion object {
        @JvmStatic
        val instance: FileClickHandlerRegistry by lazy { FileClickHandlerRegistry() }
    }
}
